Add codelab: deploying Artifactory artifacts to Cloud Foundry (#1374)
spinnaker/spinnaker.github.io: guides/tutorials/codelabs/artifactory-to-cf/index.md

---
 layout: single
 title:  "Deploy Artifacts from Artifactory to CF"
 sidebar:
   nav: guides
---

 {% include toc %}
 
> This codelab assumes that you have enabled the `artifactsRewrite` feature flag. In `~/.hal/$DEPLOYMENT/profiles/settings-local.js` (where `$DEPLOYMENT` is typically `default`), add:
>
> `window.spinnakerSettings.feature.artifactsRewrite = true;`

> In this codelab, you will deploy an artifact to Cloud Foundry via a Spinnaker pipeline that is triggered by an artifact published to a JFrog Artifactory Maven repository or by an app manifest stored in a GitHub Git repository.

Reviewer: ```suggestion
In this codelab, you will deploy an artifact to Cloud Foundry via a Spinnaker pipeline. The pipeline is triggered by an artifact published to a JFrog Artifactory Maven repository or by an app manifest stored in a GitHub repository.
```

## Prerequisites

This codelab assumes you have the following:

> 1. An Artifactory repository configured to accept a Maven artifact--the JAR for your application.

Reviewer: ```suggestion
* An Artifactory repository configured to accept a Maven artifact&mdash;the JAR for your application
```

> 1. A GitHub repository containing a manifest with which to deploy your application.

Reviewer: ```suggestion
* A GitHub repository containing a manifest with which to deploy your application
```

a. Follow the documentation about [Configuring GitHub Webhooks](https://www.spinnaker.io/setup/triggers/github/). Take note of the secret you provide to GitHub for the webhook.

## 5. Configure the Application and Pipeline

a. Create a new pipeline for your application. In the pipeline configuration, add an Artifactory trigger to the Automated Triggers, as shown below.

  {% include figure
     image_path="./artifactory-trigger.png"
  %}

b. In the Artifact Constraints dropdown for the Artifactory trigger, select "Define a new artifact..." and populate the Expected Artifact form with the details of the artifact that will be published to the Artifactory repository.

  {% include figure
     image_path="./artifactory-expected-artifact.png"
  %}

c. Add a Git trigger to the Automated Triggers, as shown below. Supply the secret you used in Step 4a.

  {% include figure
     image_path="./git-trigger.png"
  %}

d. In the Artifact Constraints dropdown for the Git trigger, select "Define a new artifact..." and populate the Expected Artifact form with the details of the manifest stored in the GitHub repository.

  {% include figure
     image_path="./git-expected-artifact.png"
  %}

> e. Add a "Deploy" stage to the pipeline. This stage will deploy to the test environment.

Reviewer: ```suggestion
e. Add a "Deploy" stage to the pipeline. This stage deploys to the test environment.
```

  {% include figure
     image_path="./pipeline-first-stage.png"
  %}

f. Add a new server group and provide details on deployment settings, the application artifact, and the manifest artifact:

  {% include figure
     image_path="./server-group.png"
  %}

g. Add a "Manual Judgment" stage. You can use this stage, or stages such as "Script" or "Canary Analysis", to verify a deployment before promoting it from the test environment.

  {% include figure
     image_path="./pipeline-second-stage.png"
  %}

Add any desired configuration, such as notifications for when the stage is awaiting judgment or is manually judged to continue or stop.

  {% include figure
     image_path="./manual-judgment-configuration.png"
  %}

> After you have validated and approved the deployment, the next stage (following this "Manual Judgment" stage) will deploy to the production environment.

Reviewer: ```suggestion
After you have validated and approved the deployment, the next stage (following this "Manual Judgment" stage) deploys to the production environment.
```

> h. Add a "Clone Server Group" stage to the pipeline. This stage will promote the deployed application to the production environment.

Reviewer: ```suggestion
h. Add a "Clone Server Group" stage to the pipeline. This stage promotes the deployed application to the production environment.
```

  {% include figure
     image_path="./pipeline-third-stage.png"
  %}

i. Add a clone configuration and provide details on deployment settings, the source server group, and the manifest:

  {% include figure
     image_path="./clone-configuration.png"
  %}

> This clone configuration will promote the exact same application to the production environment as you deployed to the test environment with the "Deploy" stage in Step 5e. (You may wish to provide a different manifest when deploying the application to the production environment. For this alternate manifest to trigger pipeline executions, you would need to create a separate pipeline that deploys to the production environment.)

Reviewer: ```suggestion
This clone configuration promotes the exact same application to the production environment as you deployed to the test environment with the "Deploy" stage in Step 5e. 

You may wish to provide a different manifest when deploying the application to the production environment. For this alternate manifest to trigger pipeline executions, you would need to create a separate pipeline that deploys to the production environment.
```

## 6. Publish the Application Artifact or Commit Configuration Changes

a. Publish the application artifact to your Artifactory repository. You can also update the application manifest and push a new commit to its GitHub repository.

> b. The published artifact or Git repository push will trigger a pipeline execution, and you should see the pipeline deploy a new server group for the application, then promote the application to staging.

Reviewer: ```suggestion
b. The published artifact or Git repository push triggers your pipeline, and you should see it deploy a new server group for the application, then promote the application to staging.
```
